### PR TITLE
Add GH org permission check

### PR DESCRIPTION
**Pull request by mic67mel**
Github permissions check on collaborators, forks and team added

- [ x ] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)

## What

Provide a check to control the access to Github repositories containing source code. 

## Why

Permission access should be done at the team level, and not by adding single collaborators outside of team membership.

## How

- test_org_permissions.py
- org_permissions.md.tmpl
- README.md

## Test

N/A

## Context

#58

### CODE REVIEW

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md]: This is misleading.  Checks don't control anything.  This check's purpose is to validate a condition is not violated.  State what is being validated.  The check doesn't care whether the repos it is looking at contain source code or not.

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md]: We should not be listing successes in check results or reports.  This is how you wind up with check results that are too big to be rendered in the remote hosting service (GH, GL BB).  The purpose of a check is to flag warnings and failures so that they can be remediated not to list everything that right in the world.

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md]: - `GithubOrgPermissionFetcher` should be a link
- Why are "all" collaborators excluded from this check?

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md:141]: add blank line above

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md:142]: add blank line above

**Review by alfinkel — CHANGES_REQUESTED** [review]
Why is this check specific only to Github?  Seems like you should be following the same pattern as the OrgCollaboratorsCheck which should theoretically work for Github or Gitlab or Bitbucket.

Pausing...

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md:137]: Surround this line with a blank line before and after it.

**Comment by mic67mel** [comment]
@alfinkel I can add the few lines of code to handle Gitlab and Bitbucket as in OrgCollaboratorsCheck , but I am not sure that the assumption that the Gitlab and Bitbucket evidence data will have same format as Github evidence, is correct. What do you think?

**Comment by alfinkel** [comment]
You are correct that the format of the evidence will likely be different.  But we knew this before too for the collaborators check.  When/if we decide to add support for Gitlab and/or Bitbucket then the format differences between the evidences across the 3 platforms will need to be handled in evidence subclasses but the logic of the check _should_ be applicable for all three.

It's not just adding a few lines of code.  It's refactoring the code to make the check more generic.  For example, renaming the class, etc...

**Comment by mic67mel** [comment]
Are you suggesting to rework OrgCollaboratorsCheck too? Since I don't see the logic to handle the 3 types, if I am not wrong.

**Comment by alfinkel** [comment]
I'm not suggesting that at all.  I'm asking you to refactor the code/docs in this PR to more closely resemble the OrgCollaboratorsCheck.  Thanks.

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md]: Remove this line to be consistent with the other check.

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md]: To be consistent with the other check...

```suggestion
  * NOTE: Only gh (Github) is currently supported by this check. Gitlab and Bitbucket support coming soon...
```

File: arboretum/permissions/README.md
Comment (alfinkel) [reply arboretum/permissions/README.md:137]: Please add a blank line after this Import statement.

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: To be consistent with the other check...

```suggestion
    """Checks for repository organization/owner permissions."""
```

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: To be consistent with the other check...

```suggestion
        return 'Repository Organization/Owner Permissions'
```

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: again...

```suggestion
                    'Repository organization permission report.'
```

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: This is wrong.  A split returns a list.  Any comparison done on L55 and L57 will never work.  You're basically introducing a dormant bug that will wake up if we ever add support for GL and BB.

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: If you handled the split correctly to get the `host` and `org_name` as is done in [the check](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/permissions/checks/test_org_collaborators.py#L57-L70) that you should be using as a guide then you could also have the `with evidences(...)` context manager at the same level as the `for org in orgs:` loop.  Unless there's a good reason to have the context manager imbedded in the `for` loop?

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: Again, I suggest you follow more closely the pattern established in the [original check](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/permissions/checks/test_org_collaborators.py#L65-L67).

File: arboretum/permissions/README.md
Comment (alfinkel) [inline arboretum/permissions/README.md]: It looks like it's direct and outside collaborators that generate the failure.  Am I missing something?

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: This looks very odd to me.  It _seems_ like `ev_direct` is a dictionary where the key is `repo` and repo is a string?

Can you provide an example of what the evidence looks like here?  It seems like you can do ` for repo, c in ev_direct.items()` instead of this nested for loop.

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: When exactly would collabs not be collabs if it gets past the `continue`?  Seems like a useless `if` statement.

**Review by alfinkel — CHANGES_REQUESTED** [review]
I think you should have a look at the [collabs check](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/permissions/checks/test_org_collaborators.py) again.  There are some deviations from it in this PR that seem problematic.  It also seems like you may not be using Python correctly in some spots.

**Comment by alfinkel** [comment]
It also looks like you're missing the notifier method so your notifiers would not work.  That should be something you test going forward.

File: arboretum/permissions/README.md
Comment (mic67mel) [reply arboretum/permissions/README.md]: Failure is genereted for direct collaborators, outside collaborators are used to determine if a direct is org member or not.

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: C is a collaborators list, I need the second loop. I will send you the link to the evidence file

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: I need the nested loop because I have to handle multiple evidence files non only one as in the other check. The evidence path key must be the evidence type(direct, outside or forks)

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: This is pretty ugly and hard to read.  If you renamed some things to get this to fit 80 char then you could rewrite this as:

```suggestion
            path = 'raw/permissions/{}_{}_{}.json'
            evs = ['direct_collaborators', 'outside_collaborators', 'forks']
            ev_paths = {ev: path.format(service, ev, url_hash) for ev in evs}
```

`evidence_paths` would now be `ev_paths` and its keys would be 'direct_collaborators', 'outside_collaborators', 'forks'  but I think this is better.

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: `raws`?  How about just `evidence`?

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [reply arboretum/permissions/checks/test_org_permissions.py]: `ev_direct` is a dictionary and you are interested in both the key (the repo) and the value (the collaborator list) so you should be using the `items()` method in the for loop when looking through `ev_direct`.

L81 should be `for repo, collab_list in ev_direct.items():`

L83 would then be `for c in collab_list:`, but my preference is for it to be `for collab in collab_list:` since I'm not really sure why all of a sudden you switched to 1 letter variable names.

**Review by alfinkel — CHANGES_REQUESTED** [review]
Pausing...

It seems like you've missed some of my comments from the previous review.  Here is [one](https://github.com/ComplianceAsCode/auditree-arboretum/pull/60/files#r622609757).  I believe there are others.  

You should apply the comment about using `.items()` whenever you step through a dictionary.  It is more pythonic and matches the code style of this repo.  There are other instances of this in your PR.

Please move your "private" methods to the end of the check class.  It's a standard and also matches the code style of this repo.

Use snake case when naming your variables.  For example `repocollabs` should be `repo_collabs`.  Apply this throughout the PR code.

File: arboretum/permissions/README.md
Comment (mic67mel) [reply arboretum/permissions/README.md]: Done

File: arboretum/permissions/README.md
Comment (mic67mel) [reply arboretum/permissions/README.md]: Done

File: arboretum/permissions/README.md
Comment (mic67mel) [reply arboretum/permissions/README.md]: Done

File: arboretum/permissions/README.md
Comment (mic67mel) [reply arboretum/permissions/README.md]: Done

File: arboretum/permissions/README.md
Comment (mic67mel) [reply arboretum/permissions/README.md]: Done

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Done

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Done

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Done

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Done

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Done

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Done

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Removed the useless statement

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Done

File: arboretum/permissions/checks/test_org_permissions.py
Comment (mic67mel) [reply arboretum/permissions/checks/test_org_permissions.py]: Done

**Comment by mic67mel** [comment]
@alfinkel All comments should be addressed.

File: arboretum/permissions/checks/test_org_permissions.py
Comment (alfinkel) [inline arboretum/permissions/checks/test_org_permissions.py]: subtitle is redundant since there is only one check.

```suggestion
        return {'body': None}
```

File: arboretum/permissions/templates/reports/permissions/org_permissions.md.tmpl
Comment (alfinkel) [inline arboretum/permissions/templates/reports/permissions/org_permissions.md.tmpl]: Add a blank line to the end of this file.

**Review by alfinkel — CHANGES_REQUESTED** [review]
Should be good to merge after this.  Feel free to squash commits.

**Review by alfinkel — APPROVED** [review]

**Pull request merged**
